src/core/service.c

```diff
@@ -5003,6 +5062,137 @@ static int service_can_clean(Unit *u, ExecCleanMask *ret) {
         return 0;
 }
 
+static int service_live_mount(Unit *u,
+                const char *src,
+                const char *dst,
+                sd_bus_message *message,
+                bool read_only,
+                bool make_file_or_directory,
+                bool is_image,
+                MountOptions *options) {
+
+        _cleanup_(pidref_done) PidRef worker = PIDREF_NULL;
+        Service *s = ASSERT_PTR(SERVICE(u));
+        const char *propagate_directory;
+        PidRef *unit_pid;
+        ExecContext *c;
+        int r;
+
+        assert(u);
+        assert(src);
+        assert(dst);
+        assert(message);
+
+        if (s->state != SERVICE_RUNNING)
+                return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EBUSY), "Service is not running");
+
+        unit_pid = unit_main_pid(u);
+        if (!pidref_is_set(unit_pid))
+                return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EBUSY), "Service is not running");
+
+        c = unit_get_exec_context(u);
+        if (!c)
+                return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EINVAL), "Cannot access unit execution context");
+
+        service_unwatch_control_pid(s);
+        s->mount_result = SERVICE_SUCCESS;
+        s->control_command = NULL;
+        s->control_command_id = _SERVICE_EXEC_COMMAND_INVALID;
+
+        r = service_arm_timer(s, /* relative= */ true, s->exec_context.timeout_mount_usec);
+        if (r < 0) {
+                log_unit_warning_errno(u, r, "Failed to install timer: %m");
+                goto fail;
+        }
+
+        propagate_directory = strjoina("/run/systemd/propagate/", u->id);
+
+        /* Given we are running from PID1, avoid doing potentially heavy I/O operations like opening images
+         * directly, and instead fork a worker process. We record the D-Bus message, so that we can reply
+         * after the operation has finished. This way callers can wait on the message and know that the new
+         * resource is available (or the operation failed) once they receive the response. */
+        r = unit_fork_helper_process(u, "(sd-mount-in-ns)", /* into_cgroup= */ false, &worker);
```

Comment:
I'm not convinced this forker process should live out of unit cgroup? All the accounting should definitely be assigned to unit cgroup, plus GetUnitByPID and such?

Reply:
See commit message - that would be a regression and cause OOM kills, so it cannot happen

Reply:
I am really not convinced by this. Let's not do this. We insert a tiny helper. We block on IO yes, but we do not process gigabytes of IO. Such a limited IO should always be fine, in particular as the implementation of reloading inserts things the same way.

Hence, please, let's not get paranoid here. Please drop this, just make this part of the cgroup of the service. We can revisit this if this really turns out to be a problem. But otherwise I am not impressed at all, this is theatre.

Reply:
@poettering It's not about IO though, It's about avoiding exactly the same issue that systemd-executor solves namely the copy-on-write nature of fork() and all that memory usage getting attributed to the target cgroup.

If you'd want it to be attributed to the target cgroup we'd need to vfork() + execve() to avoid the copy-on-write trap.

Reply:
Yeah and even with vfork I'm not sure if it would help for this case. This is not paranoia, I absolutely have real world use cases running like this, with a fixed MemoryMax, in the 100M range or so, that is very very very close to the actual peak memory usage. Dumping even 10M in that cgroup means the kernel OOM killer goes ham, and breaks the system, and this would be _hard_ breakages.

With the `into_cgroup` flag set to true:

```
# systemd-run -p PrivateTmp=yes --unit test.service -p MemoryMax=100M stress-ng --timeout 3m --vm 10 --vm-bytes 70M --vm-keep
# systemctl mount-image --mkdir test.service /usr/share/minimal_0.raw /tmp/bar
<...>
[ 1191.740618] Memory cgroup out of memory: Killed process 2023 (stress-ng-vm) total-vm:73336kB, anon-rss:9424kB, file-rss:1532kB, shmem-rss:0kB, UID:0 pgtables:104kB oom_score_adj:1000
```

This is really not something just in case, but a real hard requirement, I cannot have this stuff regress.

Reply:
if the vfork() ssue matters to you, I'd be happy if we'd move this into an external binary, too. In fact all those processes we fork off like this might benefit from that.

Reply:
Again, even that wouldn't help, as it's still multiple MBs suddenly being added to a cgroup where previously none where, so it will still cause OOM kills as shown above.